migration/idempotent/017-metric-downsampling.sql

```diff
@@ -0,0 +1,392 @@
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.downsample_counter(_schema TEXT, _table_name TEXT, _resolution INTERVAL)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+BEGIN
+    EXECUTE FORMAT(
+        'CREATE MATERIALIZED VIEW %1$I.%2$I WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS
+            SELECT
+                timezone(
+                    %3$L,
+                    public.time_bucket(%4$L, time) AT TIME ZONE %3$L + %4$L
+                ) as time,
+                series_id,
+                public.first(value, time),
+                public.last(value, time) + _prom_catalog.counter_reset_sum(array_agg(value)) last,
+                _prom_catalog.irate(array_agg(value)) irate
+            FROM prom_data.%2$I
+            GROUP BY public.time_bucket(%4$L, time), series_id WITH NO DATA
+    ', _schema, _table_name, 'UTC', _resolution::text);
+END;
+$$
+LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.downsample_summary(_schema TEXT, _table_name TEXT, _resolution INTERVAL)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+BEGIN
+    EXECUTE FORMAT(
+        'CREATE MATERIALIZED VIEW %1$I.%2$I WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS
+            SELECT
+                timezone(
+                    %3$L,
+                    public.time_bucket(%4$L, time) AT TIME ZONE %3$L + %4$L
+                ) as time,
+                series_id,
+                sum(value) as sum,
+                count(value) as count
+            FROM prom_data.%2$I
+            GROUP BY public.time_bucket(%4$L, time), series_id WITH NO DATA
+    ', _schema, _table_name, 'UTC', _resolution::text);
+END;
+$$
+LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.add_compression_clause_to_downsample_view(_schema TEXT, _table_name TEXT)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+BEGIN
+    EXECUTE FORMAT($sql$
+        ALTER MATERIALIZED VIEW %1$I.%2$I SET (timescaledb.compress = true)
+    $sql$, _schema, _table_name);
+END;
+$$
+LANGUAGE PLPGSQL;
+
+-- This creates a view that acts as a interface between connector and the downsampled data while querying.
+-- We need this primarily for PromQL querying. Downsampled data does not contain the default column "value"
+-- that connector's SQL query needs, when querying a downsampling view. Hence, to avoid complexity of code
+-- on the connector side ("value" columns differs depending on metric type), we create a view that internally
+-- maps to the right column of downsampled data.
+CREATE OR REPLACE FUNCTION _prom_catalog.create_default_downsampling_query_view(_schema TEXT, _table_name TEXT, _default_column TEXT)
+RETURNS VOID
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+DECLARE
+    -- We prefix with q_ to avoid collision with names of materialized views. This prefix can be added
+    -- on the connector side.
+    query TEXT := 'CREATE OR REPLACE VIEW %1$I.q_%2$I AS SELECT time, series_id, %3$s AS value FROM %1$I.%2$I';
```

Comment:
See commit [Adds support to query default downsampling views in absence of `__column__`](https://github.com/timescale/promscale/pull/1808/commits/7464174d04281b2e81652fa8c1c6c08b13378075) for understanding the use of this view.